tools/githooks/pre-push

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python3
+# Copyright 2013 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+'''
+Runs the pre-push githooks.
+'''
+
+import os
+import subprocess
+import sys
+
+
+SRC_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
+FLUTTER_DIR = os.path.join(SRC_ROOT, 'flutter')
+DART_BIN = os.path.join(SRC_ROOT, 'third_party', 'dart', 'tools', 'sdks', 'dart-sdk', 'bin')
```

Comment:
I have vague memories of this causing subtle problems in the past, but I can't quite remember what. Something about versions being slightly off from what people would normally expect to work with, as well as confusion about which Dart binary was actually getting used and for what. @chinmaygarde may remember.

Overall the code in here looks good, but I'm wondering if we're making things overly complicated here - we have gclient invoke a python script, which tells git to invoke a python script, which invokes a Dart program, which invokes some bash scripts that we'll someday refactor to either python or Dart.

Why not just make everything python?

Reply:
> I have vague memories of this causing subtle problems in the past, but I can't quite remember what. Something about versions being slightly off from what people would normally expect to work with, as well as confusion about which Dart binary was actually getting used and for what. @chinmaygarde may remember.

Git hooks only work with git version >= 2.9. Is that what you mean?

If you mean the prebuilt Dart SDK from CIPD, we already use it pretty consistently to run standalone Dart programs in the tree.

> Overall the code in here looks good, but I'm wondering if we're making things overly complicated here - we have gclient invoke a python script, which tells git to invoke a python script, which invokes a Dart program, which invokes some bash scripts that we'll someday refactor to either python or Dart.

It's not quite that bad. To set the git config, `gclient sync` invokes a python script. Then separately, on a `git push`, `git` invokes the python script, which invokes the Dart program (which invokes a shell script, which invokes another Dart program, which invokes e.g. clang-tidy in the linter case.)

The part in parens can be simplified by refactoring the `ci/bin/lint.dart` and `ci/bin/format.dart` scripts into proper Dart packages. I've started doing that here: https://github.com/flutter/engine/pull/26722. Then it's just python -> Dart -> clang-tidy. The python part is kind of mandatory due to the need to have a `#!` at the top for the git hook.

> Why not just make everything python?

My first pass at the `pre-push` hook was python. There were two problems:
1. I was thinking about how I was going to extend it to analyze only a minimal set of Dart code on changes to .dart files. I found myself reimplementing in python some logic that we already had implemented in Dart in `ci/bin/format.dart` and `ci/bin/lint.dart`. So that didn't seem like the right approach.
2. I was writing a lot of python, and I wanted to be able to lint the code and write tests for it. Currently that's easy in Dart, but we don't have any setup for that for python. We should add it for python, probably, but given (1) it seemed like the right thing was to just use Dart.

Reply:
My concerns weren't about the hooks setting for git, but more about whether we can trust the Dart binary. It's also possible that I'm mixing something up here - I think the problem may have come from trying to invoke Dart directly as a gclient hook (not a git hook). At any rate, if we're using it successfully elsewhere we should be fine.

This change LGTM